FinalLocalVariable: NPE in records

Check documentation: https://checkstyle.org/config_coding.html#FinalLocalVariable

problem was detected during https://github.com/checkstyle/contribution/commit/529fc836f88e0f88148ab7760a5ec7ba4d100715
`**/resources-noncompilable/**/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityRecords.java` is name of file in our repo that cause problem.

```
$ cat MyRecord2.java 
//non-compiled with javac: Compilable with Java14

public record MyRecord2(boolean t, boolean f) {
    public MyRecord2 { // violation
        int a = 1;
        int b = 1;
        int c = 1;
        int d = 1;
        int n = 8;
        if (a == 1) { // 2, if
        } else if (a == b // 3, if
                && a == c) { // 4, && operator
            if (c == 2) { // 5, if
            }
        } else if (a == d) { // 6, if
            try {
            } catch (Exception e) { // 7, catch
            }
        } else {
            switch (n) {
                case 1: // 8, case
                    break;
                case 2: // 9, case
                    break;
                case 3: // 10, case
                    break;
                default:
                    break;
            }
        }
        d = a < 0 ? -1 : 1; // 11, ternary operator
    }
}

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
       <module name="FinalLocalVariable"/>
    </module>
</module>

$ java -jar checkstyle-8.36.2-all.jar -c config.xml MyRecord2.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing MyRecord2.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:311)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.findFirstUpperNamedBlock(FinalLocalVariableCheck.java:781)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.isSameVariables(FinalLocalVariableCheck.java:800)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.access$700(FinalLocalVariableCheck.java:142)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck$ScopeData.findFinalVariableCandidateForAst(FinalLocalVariableCheck.java:863)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.getFinalCandidate(FinalLocalVariableCheck.java:444)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.visitToken(FinalLocalVariableCheck.java:304)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:340)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:451)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:278)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:151)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:298)
	... 5 more
Checkstyle ends with 1 errors
```

after fix we should remove this exclude from contribution repo.

It's interesting that a null pointer exception occurred. It shouldn't.
When I look at the exception stack trace:

```
Caused by: java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.findFirstUpperNamedBlock(FinalLocalVariableCheck.java:781)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.isSameVariables(FinalLocalVariableCheck.java:800)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.access$700(FinalLocalVariableCheck.java:142)
	at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck$ScopeData.findFinalVariableCandidateForAst(FinalLocalVariableCheck.java:863)
.
.
.
```
Basically `findFinalVariableCandidateForAst` called `isSameVariables`:
```
        public Optional<FinalVariableCandidate> findFinalVariableCandidateForAst(DetailAST ast) {
            Optional<FinalVariableCandidate> result = Optional.empty();
            DetailAST storedVariable = null;
            final Optional<FinalVariableCandidate> candidate =
                Optional.ofNullable(scope.get(ast.getText()));
            if (candidate.isPresent()) {
                storedVariable = candidate.get().variableIdent;
            }
            if (storedVariable != null && isSameVariables(storedVariable, ast)) { // RIGHT HERE
                result = candidate;
            }
            return result;
        }
```
`storedVariable` is checked != null. So it shouldn't be equal to null.
Then `isSameVariables` called `findFirstUpperNamedBlock`:
```
private static boolean isSameVariables(DetailAST ast1, DetailAST ast2) {
        final DetailAST classOrMethodOfAst1 =
            findFirstUpperNamedBlock(ast1); // RIGHT HERE
        final DetailAST classOrMethodOfAst2 =
            findFirstUpperNamedBlock(ast2);
        return classOrMethodOfAst1 == classOrMethodOfAst2 && ast1.getText().equals(ast2.getText());
    }
```
`ast1` is `storedVariable`. So it shouldn't be null either.
Then it crashed at `findFirstUpperNamedBlock` due to `null pointer exception`:
```
private static DetailAST findFirstUpperNamedBlock(DetailAST ast) {
        DetailAST astTraverse = ast;
        while (astTraverse.getType() != TokenTypes.METHOD_DEF // CRASHED HERE
                && astTraverse.getType() != TokenTypes.CLASS_DEF
                && astTraverse.getType() != TokenTypes.ENUM_DEF
                && astTraverse.getType() != TokenTypes.CTOR_DEF
                && !ScopeUtil.isClassFieldDef(astTraverse)) {
            astTraverse = astTraverse.getParent();
        }
        return astTraverse;
    }
```
But `astTraverse` shouldn't be null. (I haven't run the debugger. I will do that tomorrow morning).
Did I miss something here? The only thing I didn't understand is this:
```
at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.access$700(FinalLocalVariableCheck.java:142)
```
It stands between `findFinalVariableCandidateForAst` and `isSameVariables` in the exception stack trace. So maybe it does something that I don't know about?

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java
@@ -33,6 +33,7 @@ import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
 import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
  * <p>
@@ -778,10 +779,8 @@ public class FinalLocalVariableCheck extends AbstractCheck {
      */
     private static DetailAST findFirstUpperNamedBlock(DetailAST ast) {
         DetailAST astTraverse = ast;
-        while (astTraverse.getType() != TokenTypes.METHOD_DEF
-                && astTraverse.getType() != TokenTypes.CLASS_DEF
-                && astTraverse.getType() != TokenTypes.ENUM_DEF
-                && astTraverse.getType() != TokenTypes.CTOR_DEF
+        while (!TokenUtil.isOfType(astTraverse, TokenTypes.METHOD_DEF, TokenTypes.CLASS_DEF,
+                TokenTypes.ENUM_DEF, TokenTypes.CTOR_DEF, TokenTypes.COMPACT_CTOR_DEF)
                 && !ScopeUtil.isClassFieldDef(astTraverse)) {
             astTraverse = astTraverse.getParent();
         }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
@@ -87,6 +87,17 @@ public class FinalLocalVariableCheckTest
     }
 
     @Test
+    public void testRecordsInput() throws Exception {
+        final DefaultConfiguration checkConfig =
+            createModuleConfig(FinalLocalVariableCheck.class);
+        final String[] expected = {
+            "15:17: " + getCheckMessage(MSG_KEY, "b"),
+        };
+        verify(checkConfig,
+            getNonCompilablePath("InputFinalLocalVariableCheckRecords.java"), expected);
+    }
+
+    @Test
     public void testParameter() throws Exception {
         final DefaultConfiguration checkConfig =
             createModuleConfig(FinalLocalVariableCheck.class);

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableCheckRecords.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableCheckRecords.java
@@ -1,0 +1,19 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.coding.finallocalvariable;
+
+/* Config:
+ * default
+ */
+public record InputFinalLocalVariableCheckRecords(boolean t, boolean f) {
+    public InputFinalLocalVariableCheckRecords {
+        int a = 0; // ok
+        a = 1;
+    }
+
+    record bad(int i) {
+        public bad {
+            int b = 0; // violation
+        }
+    }
+}
+

